### ISSUE TITLE
Request for inclusion of external package "photospline"

### ISSUE THREAD

**Issue by lgray** [cms-sw/cmsdist#145]
Hi,

If it is possible, I'd like to include this external package for fast multi-dimensional spline fitting in CMSSW. I need it to provide cross section and shape-morphing information for a combined HIG-SMP fit (Higgs + Multiboson).
I've been able to get it to compile within the standard CMSSW environment, but can be a little painful to setup since it uses cmake and it would be nice to have it properly integrated and recognized as a scram tool.

You can find the source code in the pre-print tarball at the arxiv:
http://arxiv.org/abs/1301.2184

Its dependencies are:
CMake (>= 2.8): http://www.cmake.org
cfitsio (>= 3.1): http://heasarc.gsfc.nasa.gov/fitsio/
gsl (>= 1.14): http://www.gnu.org/software/gsl/
SuiteSparse (>= 3.5): http://www.cise.ufl.edu/research/sparse/
PyFITS (>= 3.1): http://www.stsci.edu/institute/software_hardware/pyfits
Numpy (>= 1.3): http://www.numpy.org
Python (>= 2.5): http://www.python.org

Finally it needs an optimized BLAS and LAPACK.
The authors recommend: http://www.tacc.utexas.edu/tacc-projects/gotoblas2

For configuration details there's a thorough description in the source tarball under photospline/README.txt.

Please let me know if I need to edit any packages to compile with scram or if there is anything else I can do help this along.

Thanks!

**Comment by davidlt** [comment]
@lgray , this is not a small request. It requires a number of new packages included into CMSSW. Tomorrow there will be a [DORP](https://indico.cern.ch/conferenceDisplay.py?confId=254370) meeting, where we discuss new requests. You need to convince people that it's must-have externals for your feature and the benefits it will brings.   Also you have to make sure that it's currently impossible or complicated enough to achieve the same using the current external packages or CMSSW code.

Currently we have CMake, Numpy, gsl and Python available in CMSDIST. Thus PyFITS, SuiteSparse, cfitsio, BLAS, photospline are the new externals.

We do have LAPACK from http://www.netlib.org/ which is not good for photospline according to README and should be avoided.

And for BLAS it's GotoBLAS 2 (Linux) and Accelerate.framework (Darwin).

**Comment by lgray** [comment]
@davidlt I haven't been able to make it to an offline meeting yet either due to travel or needing to go to other meetings. I will try to make it to the next one, which is again this Friday?

FYI:
- An optimized LAPACK is provided already with GotoBLAS2 and Accelerate.framework.
- The netlib reference BLAS is also already in CMSSW, but again this is disfavored.

A quick summary of my points would be that: 
- Using the current tools available in CMSSW, evaluation and fitting of high-dimensional splines to data is either computationally very slow or impossible due to memory limitations. Standard interpolation techniques are memory hogs and require many evaluations of splines. This worsens at higher dimensionality. 
- This package solves both of those issues using computationally efficient evaluation techniques and highly optimized linear algebra and sparse matrix packages. 
- I need to be able to package this functionality with the combine tool to be able to do the analysis that I would like to do. This requires packaging with CMSSW.
- Having highly optimized linear algebra and spline fitting packages will be of use to more than just me, since modelling high-dimensional data in a compact and efficient way is a problem that is common to other analyses and even reconstruction techniques.

**Comment by davidlt** [comment]
@lgray , [next meeting](https://twiki.cern.ch/twiki/bin/view/CMS/OfflineComputingPlanningMeeting20130802) is tomorrow.

**Comment by lgray** [comment]
Hey, I will be on vacation starting Monday, so I will get back to this after I return on the 14th of August.
I'll provide to you all the necessary recipes for installing these packages.

Is there any specific area you need them to be installed to or are install scripts that simply work within a CMSSW environment enough?

**Comment by lgray** [comment]
@davidlt You can find a working annotated install script at: /afs/cern.ch/user/l/lgray/work/public/install_photospline.sh

Instructions:

``` bash
cmsrel CMSSW_7_0_0_pre2 && cd CMSSW_7_0_0_pre2/src
cmsenv
cp /afs/cern.ch/user/l/lgray/work/public/install_photospline.sh ./
source install_photospline.sh
# you should then be able to run without any issue
python photospline/examples/splinefit_1d.py
python photospline/examples/pseudointerpolation.py
```

I suspect many of the issues that I ran into will be fixed when everything upstream of the new packages is install correctly.

The updates to GSL and matplotlib are important and the need to be installed after and linked to OpenBLAS.
We can probably drop the reference BLAS implementation that's in CMSSW right now.

I'll talk with the photospline authors about not being able to import spglam (right now it fails over to the very very slow python implementation). This shouldn't be too hard to fix.

CMake doesn't need to be a shipped package in the end, it is simply there to build photospline, so it is not necessary to keep it around after building is done.

Disk usage after removing tar balls and cmake is 137 MB, more than half of that is the new matplotlib which is already part of CMSSW.
The next largest package is OpenBLAS which comes in at 34 MB and should likely replace the current reference BLAS and LAPACK that we have. The interfaces are the required to be the same as netlib and OpenBLAS will be significantly faster for low-level BLAS routines. 

The total amount of truly 'new' stuff being shipped weights in at ~20 MB.

**Comment by lgray** [comment]
@davidlt The script is now double checked and completely ready, please give it a spin.

**Comment by davidlt** [comment]
@lgray , I am working on it. I'll ping the thread once it's ready.

**Comment by lgray** [comment]
@davidlt 

Hmm, I've tracked down the issue with spglam:

```
/.../spglam.so: undefined symbol: clock_gettime
```

This is either part of libc or librt... The weird thing being that the library compiles without issue and fails to run, and these are standard system libraries.

Any ideas?

**Comment by davidlt** [comment]
Most likely missing `-lrt` or in a wrong position (position dependent arguments). You can also enforce linker not to allow any undefined symbols. Some packages tends to build with undefined references and they expect that target application will load it during run-time.

@lgray , _spglam_ comes from which package?

**Comment by lgray** [comment]
@davidlt it's part of photospline. spglam is the python extension for using the C-based spline fitter (needed for high-dimensional fits)

**Comment by lgray** [comment]
@davidlt got it working, had to make sure it was using the correct python version and then force it to load librt and now it works (and is super fast). 
All that needs changing is the configuration of cmake for photospline. 
I've updated /afs/cern.ch/user/l/lgray/work/public/install_photospline.sh to reflect this.

**Comment by lgray** [comment]
@davidlt Hey, just checking in. Encountered any problems or otherwise concerning issues?

**Comment by davidlt** [comment]
@lgray Currently I had time to do half of of RPMs. I stopped with `OpenBLAS`, at first glance it tries to build micro-architeture optimized version, which is no good for us. We run on various INTEL and AMD x86_64 CPUs, also ARMv7-A. I need to see if it's possible to get a generic, not micro-architecture optimized build.

**Comment by lgray** [comment]
@davidlt There is the option in OpenBLAS `DYNAMIC_ARCH=1` which builds all optimized libraries and dynamically chooses the best at run time... Of course this may explode the size of the package. I will check and let you know.

**Comment by lgray** [comment]
@davidlt checked it, compiles with no issue. 
The size of OpenBLAS goes from 34MB -> 85MB, but now dynamically determines which architecture to use (anything that's x86/x86_64 gets linked into the dynamic library).
I imagine that ARM has to be cross-compiled for, there is a 'generic' `TARGET=GENERIC` option, but it is horribly slow to use on x86_64 machines. I don't have access to an ARM machine to check it, if you point me to one I can use can I try it out.

Bottom line: using this feature solves most of this problem.

I've updated /afs/cern.ch/user/l/lgray/work/public/install_photospline.sh to include the dynamic architecture feature.

**Comment by lgray** [comment]
@davidlt 

Tried to build OpenBLAS on an arm7 system (armcms01). 
Doesn't work out of the gate in 'generic' mode, states that "This arch/CPU is not supported by OpenBLAS.".

However, it looks like it's pretty easy to build up the necessary bits to make it at least compile the generic code if it is on an arm system. Just some makefile and C system information stuff.

I'll keep you informed.

**Comment by lgray** [comment]
@davidlt 

Ok, got some of it working. The present roadblock is that there is no fortran compiler on the armcms machines.

**Comment by ktf** [comment]
Ciao Lindsey,

I finally managed to find some time for this. I've built and uploaded photospline and py2-pyfits for slc5_amd64_gcc481.

They are not yet in configuration, however you can set them up by doing:

```
scram setup /afs/cern.ch/cms/slc5_amd64_gcc481/external/photospline-toolfile/1.0/etc/scram.d/photospline.xml
scram setup /afs/.cern.ch/cms/slc5_amd64_gcc481/external/py2-pyfits-toolfile/1.0-cms/etc/scram.d/py2-pyfits.xml
```

I ran the tests you suggested and they were fine. If you are happy as well, we can start including them in the configuration (BTW, do you have something more beefy we can try it out?).

As discussed we can then try out a few things:
- A drop in openblas replacement.
- The TBB version.
- The CUDA version.

probably for next week though...
